### PR TITLE
Typed Persistence: fix snapshot deletion algorithm, #26584

### PR DESCRIPTION
**Pull request by patriknw**
* Subtraction of sequenceNr (keepNSnapshots * snapshotEveryNEvents) was missing when
  calling internalDeleteSnapshots directly from SaveSnapshotSuccess, i.e. when not
  triggered via delete events.
* Improved the test
* Also found that the responses for deletion of events and snapshots were not
  handled in all 3 running phases. Since they are performed in the background
  they may be received in all.

Refs #26584

### CODE REVIEW

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/EventSourcedSignal.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/EventSourcedSignal.scala]: I like this much better, for consistency.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala]: This is curious because `math.max` is in the code this uses
https://github.com/akka/akka/blob/master/akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala#L159

So this change would mean
```scala
protected def internalDeleteSnapshots(toSequenceNr: Long): Unit = {
    val deleteTo = toSequenceNr - 1
    val func = math.max(0, toSequenceNr - (keepNSnapshots * snapshotEveryNEvents))
    val deleteFrom = math.max(0, func)
    val snapshotCriteria = SnapshotSelectionCriteria(minSequenceNr = deleteFrom, maxSequenceNr = deleteTo)
  ....
}
```

From what I think is currently
```scala
protected def internalDeleteSnapshots(toSequenceNr: Long): Unit = {
    val deleteTo = toSequenceNr - 1
    val func = toSequenceNr - (keepNSnapshots * snapshotEveryNEvents)
    val deleteFrom = math.max(0, func)
    val snapshotCriteria = SnapshotSelectionCriteria(minSequenceNr = deleteFrom, maxSequenceNr = deleteTo)
  ....
}
```

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala]: 👍

**Review by helena — APPROVED** [review]
LGTM If you're sure about the math.max move. I'll defer to you.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala]: Not sure how I missed this before but this confuses me: we call `Retention.toSequenceNumber(toSequenceNr)` and catch that in `fromSequenceNr`, so the `toSequenceNumber` method is in fact giving us a safe `from`? Can naming make this more clear?


Also, would it be a bug if we ever passed 0 as `toSequenceNr` here? Should we fail rather than silently swallow it?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala
Comment (johanandren) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala]: Maybe remove the scaladoc on this method that describes A. more logic than this method and B. the exact logic that is in this method?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: Does the snapshotter only ever respond on delete and reach here, no chance we could get a save-ok?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (johanandren) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: Feels a bit weird with a catch-all and then assume response is delete event response (and then another catchall in the end of that) - setting ourselves up for foot shooting?

**Review by johanandren — APPROVED** [review]
Mostly looks good, some improvement requests but you can decide what you think about them.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/Retention.scala]: I'm not sure I understand what you mean? I have changed that `internalDeleteSnapshots` in this PR.

I placed the `max(0` here because otherwise I would have to do that on several places. Negative sequenceNr doesn't make much sense. First event has nr 1.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/ExternalInteractions.scala]: > fromSequenceNr

I tried to explain in the comment above. It's just a way to create a sliding window.
I agree that the naming of that `Retention.toSequenceNumber` is confusing since it's used both for finding the real `to`, but also to find the `from`. I'll add that to ticket https://github.com/akka/akka/issues/26545#issuecomment-475629827

> Also, would it be a bug if we ever passed 0 as toSequenceNr here? Should we fail rather than silently swallow it?

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (patriknw) [reply akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: Alternative would be to compose partial functions, not sure that is any better. This `onDeleteEventsJournalResponse` is shared and used from several places.

I could try to make `onDeleteEventsJournalResponse` a Behavior and compose with `orElse`. I'd rather do that in a separate follow up https://github.com/akka/akka/issues/26597

**Review by patriknw — COMMENTED** [review]
Thanks for feedback, I have created follow up tickets to not delay with more code changes to this "critical section"

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: I see, I missed this happening here, in my earlier comment, versus the trait.

File: akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala
Comment (helena) [inline akka-persistence-typed/src/main/scala/akka/persistence/typed/internal/Running.scala]: 🤘

**Pull request merged**
